overlord/devicestate/handlers_systems.go

```diff
@@ -121,6 +132,121 @@ func purgeNewSystemSnapFiles(logfile string) error {
 	return s.Err()
 }
 
+func snapsUniqueToRecoverySystem(target string, systems []*System) ([]*seed.Snap, error) {
+	requiredByOtherSystems := make(map[string]bool)
+	for _, sys := range systems {
+		if sys.Label == target {
+			continue
+		}
+
+		sd, err := seed.Open(dirs.SnapSeedDir, sys.Label)
+		if err != nil {
+			return nil, err
+		}
+
+		if err := sd.LoadAssertions(nil, func(*asserts.Batch) error {
+			return nil
+		}); err != nil {
+			return nil, err
+		}
+
+		if err := sd.LoadMeta(seed.AllModes, nil, timings.New(nil)); err != nil {
+			return nil, err
+		}
+
+		err = sd.Iter(func(sn *seed.Snap) error {
+			requiredByOtherSystems[sn.Path] = true
+			return nil
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	targetSeed, err := seed.Open(dirs.SnapSeedDir, target)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := targetSeed.LoadAssertions(nil, func(*asserts.Batch) error {
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	if err := targetSeed.LoadMeta(seed.AllModes, nil, timings.New(nil)); err != nil {
+		return nil, err
+	}
+
+	var uniqueToTarget []*seed.Snap
+	err = targetSeed.Iter(func(sn *seed.Snap) error {
+		if !requiredByOtherSystems[sn.Path] {
+			uniqueToTarget = append(uniqueToTarget, sn)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return uniqueToTarget, nil
+}
+
+func (m *DeviceManager) doRemoveRecoverySystem(t *state.Task, _ *tomb.Tomb) error {
+	systems, err := m.Systems()
+	if err != nil {
+		return fmt.Errorf("cannot get recovery systems: %w", err)
+	}
+
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	setup, err := taskRemoveRecoverySystemSetup(t)
+	if err != nil {
+		return err
+	}
+
+	for _, sys := range systems {
+		if sys.Label == setup.Label && sys.Current {
+			return fmt.Errorf("cannot remove current recovery system: %q", setup.Label)
+		}
+	}
```

Comment:
@bboozzoo that check for preventing the removal of the current system is right here.